### DISCUSSION TITLE
Is there a way to check for response's status code when querying using urql client?

### DISCUSSION THREAD

**Discussion by xoldd** [urql-graphql/urql#2802] [Q&A] [answered]
There's a graphql server implementation for fastify called mercurius. It has an error class which is used to throw and return errors inside resolvers. That error class also has an option to pass in status code for the response. I wanted to use this feature to check for the status code instead of checking for errors field on the response. How to do this with urql client?

**Comment by kitten** [comment, marked as answer]
On `CombinedError` there's an untyped `response` property from which you can retrieve the status code.

As to why `urql` doesn't highlight this, we don't make many assumptions on an HTTP transport being used.

However, I wouldn't at all recommend relying on this and using the status code, except maybe for auth. But there's a better way either way.

Most errors on resolvers aren't exclusive. Meaning, you can get multiple errors in a single GraphQL response. This is important because a request can be partially fulfilled by a GraphQL API, since the schema may fail only on optional (nullable) fields, potentially deep in the response.

This is why a GraphQL Error by the spec can hold an `extensions` property. You can place an error code on there instead and in many cases it'll be more reliable to use since requests can partially succeed.

This is also useful since the response shape will then never change, plus you'll be able to differentiate GraphQL from non-GraphQL JSON responses.

http://spec.graphql.org/October2021/#sel-GAPHRPZCAACCzGg0F

**Reply by xoldd** [reply]
> On `CombinedError` there's an untyped `response` property from which you can retrieve the status code.
> 
> As to why `urql` doesn't highlight this, we don't make many assumptions on an HTTP transport being used.
> 
> However, I wouldn't at all recommend relying on this and using the status code, except maybe for auth. But there's a better way either way.
> 
> Most errors on resolvers aren't exclusive. Meaning, you can get multiple errors in a single GraphQL response. This is important because a request can be partially fulfilled by a GraphQL API, since the schema may fail only on optional (nullable) fields, potentially deep in the response.
> 
> This is why a GraphQL Error by the spec can hold an `extensions` property. You can place an error code on there instead and in many cases it'll be more reliable to use since requests can partially succeed.
> 
> This is also useful since the response shape will then never change, plus you'll be able to differentiate GraphQL from non-GraphQL JSON responses.
> 
> http://spec.graphql.org/October2021/#sel-GAPHRPZCAACCzGg0F

So, you're suggesting that I rely on graphqlErrors list instead of the status code. I'm making use of next-urql package to fetch data inside getServerSideProps. So, if I had to throw a 404 not found error I should iterate over the graphqlErrors list and check for the error that has extension code or information conveying the resource is not found. Is that correct?

**Reply by kitten** [reply]
Well, yes and no. A 404 error wouldn't be used at all for a GraphQL over HTTP endpoint as it wouldn't make much sense in most cases.

So in principle you're right. You can and should check the list of errors that's present, when you need to, for an error and check it's code. But 404 is unlikely to actually be used.

The reason for that is that if a resource isn't found that's likely modelled on the schema itself. I wouldn't expect a GraphQL schema to be modelled to error in common cases of operation. Commonly even, resource not found is often just modelled as a `null` result value.

I don't really want to present another alternative to give you too much information at once, but here's a full overview of all and another approach: https://boopathi.blog/modeling-errors-in-graphql

Basically to summarise:
- HTTP status error codes are used with GraphQL occasionally, but they're of course ambiguous when the execution of GraphQL hasn't failed completely, as it's often then unclear which resource they're for
- Extension codes on the GraphQL errors are useful since they don't force you to give the entire HTTP response a status and are specifically great for individual resolver exceptions. So when you're accessing more resolvers that may error, they're a great choice. However, they're not typed
- When a resource isn't found, sometimes just a `null` result in the schema is preferable
- In the case of mutations and some special cases, it may be preferable to model an error case in the schema entirely because you're explicitly modelling (and typing) error behaviour

I hope that gives you a comprehensive overview ✌️🙌

**Reply by xoldd** [reply]
Got it. Thanks.
